lib/evmone/instructions.hpp

```diff
@@ -173,13 +173,33 @@ inline void signextend(ExecutionState& state) noexcept
     const auto ext = state.stack.pop();
     auto& x = state.stack.top();
 
-    if (ext < 31)
+    if (ext < 31)  // For 31 we also don't need to do anything.
     {
-        auto sign_bit = static_cast<int>(ext) * 8 + 7;
-        auto sign_mask = uint256{1} << sign_bit;
-        auto value_mask = sign_mask - 1;
-        auto is_neg = (x & sign_mask) != 0;
-        x = is_neg ? x | ~value_mask : x & value_mask;
+        const auto e = ext[0];  // uint256 -> uint64.
+        const auto sign_word_index =
+            static_cast<size_t>(e / sizeof(e));      // Index of the word with the sign bit.
+        const auto sign_byte_index = e % sizeof(e);  // Index of the sign byte in the sign word.
+        const auto sign_word = x[sign_word_index];
+
+        const auto sign_byte_offset = sign_byte_index * 8;
+        const auto sign_byte = sign_word >> sign_byte_offset;  // Move sign byte to position 0.
+
+        // Sign-extend the "sign" byte and move it to the right position. Value bits are zeros.
+        const auto sext_byte = static_cast<uint64_t>(int64_t{static_cast<int8_t>(sign_byte)});
+        const auto sext = sext_byte << sign_byte_offset;
+
+        const auto sign_mask = ~uint64_t{0} << sign_byte_offset;
+        const auto value = sign_word & ~sign_mask;  // Reset extended bytes.
```

Comment:
maybe this small simplification possible?
```suggestion
        const auto sign_mask = (uint64_t{1} << sign_byte_offset) - 1;
        const auto value = sign_word & sign_mask;  // Reset extended bytes.
```

Reply:
This was the original form, but I changed this because the compiler is converting back to `x & ~(-1 << s)` (see pure -O3 version).

Things get fancy with BMI and BMI2 extensions. In BMI there is `andn` instruction (x & ~y) (GCC uses it) and `bextr` (clang).
In BMI2 these two lines are single `bzhi` instruction.

https://godbolt.org/z/K8GE9b5bW